Run2_2016/src/bbbb_nonresonant.cc

```diff
@@ -0,0 +1,176 @@
+#include "CombineHarvester/CombineTools/interface/Systematics.h"
+#include "CombineHarvester/CombineTools/interface/CardWriter.h"
+#include "CombineHarvester/CombinePdfs/interface/MorphFunctions.h"
+#include "HHStatAnalysis/Run2_2016/interface/bbbb_nonresonant.h"
+#include "HHStatAnalysis/Core/interface/exception.h"
+#include "HHStatAnalysis/Core/interface/Tools.h"
+#include "HHStatAnalysis/Core/interface/TextIO.h"
+#include "HHStatAnalysis/Core/interface/RootExt.h"
+#include "HHStatAnalysis/Run2_2016/interface/CommonUncertainties.h"
+
+namespace hh_analysis {
+namespace stat_models {
+namespace Run2_2016 {
+
+const StatModel::v_str bbbb_nonresonant::ana_name = { "HHbbbb" };
+const StatModel::v_str bbbb_nonresonant::eras = { "13TeV" };
+const StatModel::v_str bbbb_nonresonant::bkg_processes = { "bkg_hem_mix" };
+
+bbbb_nonresonant::bbbb_nonresonant(const StatModelDescriptor& _desc) :
+    StatModel(_desc), signal_processes({ desc.signal_process })
+{
+}
+
+std::string bbbb_nonresonant::NumToName(double x)
+{
+    std::ostringstream ss;
+    ss << x;
+    std::string str = ss.str();
+    std::replace(str.begin(), str.end(), '-', 'm');
+    std::replace(str.begin(), str.end(), '.', 'p');
+    return str;
+}
+
+bbbb_nonresonant::ShapeNameRule bbbb_nonresonant::BackgroundShapeNameRule() const
+{
+    std::pair<std::string, std::string> result;
+    std::ostringstream ss;
+    ss << "$PROCESS";
+    result.first = ss.str();
+    ss << "_$SYSTEMATIC";
+    result.second = ss.str();
+    return result;
+}
+
+bbbb_nonresonant::ShapeNameRule bbbb_nonresonant::SignalShapeNameRule(double /*point_value*/) const
+{
+    std::pair<std::string, std::string> result;
+    std::ostringstream ss;
+//    ss << "$BIN/$PROCESS_" << desc.signal_point_prefix << "_" << NumToName(point_value);
+    ss << "$PROCESS";
+    result.first = ss.str();
+    ss << "_$SYSTEMATIC";
+    result.second = ss.str();
+    return result;
+}
+
+void bbbb_nonresonant::ExtractShapes(ch::CombineHarvester& ch, const std::string& shapes_file, bool is_signal) const
+{
+    if(is_signal) {
+        for(double x : desc.grid_x) {
+            const auto& shape_name_rules = SignalShapeNameRule(x);
+            ch.cp().process(signal_processes).mass({ToString(x)})
+              .ExtractShapes(shapes_file, shape_name_rules.first, shape_name_rules.second);
+        }
+    } else {
+        const auto& shape_name_rules = BackgroundShapeNameRule();
+        ch.cp().process(bkg_processes).ExtractShapes(shapes_file, shape_name_rules.first, shape_name_rules.second);
+    }
+}
+
+ch::Categories bbbb_nonresonant::GetChannelCategories(const std::string& channel)
+{
+    ch::Categories ch_categories;
+    for(size_t n = 0; n < desc.categories.size(); ++n) {
+        const std::string cat_name = desc.categories.at(n);
+        const std::string name = boost::str(boost::format("%1%_%2%") % channel % cat_name);
+        ch_categories.push_back({n, name});
+    }
+    return ch_categories;
+}
```

Comment:
ok, but I'm not sure what would be wrong with function overloading in this case, instead of plain duplication...

but ok, point taken, let's go with this for the time being